File: leetcode/add-two-numbers/cpp/ans_01_01.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    typedef long long Num;
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        return write(read(l1) + read(l2));
    }
    Num read(ListNode *l)
    {
        Num res = 0;
        Num i = 0;
        while (l != nullptr)
        {
            res += l->val * pow(10, i);
            l = l->next;
            i++;
        }
        return res;
    }
    ListNode *write(Num num)
    {
        ListNode *res = new ListNode();
        ListNode *cur = res;
        while (num != 0)
        {
            cur->val = num % 10;
            num /= 10;
            if (num != 0)
            {
                cur->next = new ListNode();
                cur = cur->next;
            }
        }
        return res;
    }
};
// @lc code=end

/**
 * Line 28: Char 17: runtime error: 1e+30 is outside the range of representable values of type 'long long' (solution.cpp)
 * SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior prog_joined.cpp:37:17
 *
 * 容易 overflow, 数字太大了
 */

File: leetcode/median-of-two-sorted-arrays/py/ans_01_01.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#

# @lc code=start
import bisect


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        if m < n:
            nums1, nums2 = nums2, nums1
            m, n = n, m
        for num in nums2:
            bisect.insort(nums1, num)
        mid = (m + n) // 2
        if (m + n) % 2 == 0:
            return (nums1[mid] + nums1[mid - 1]) / 2
        else:
            return nums1[mid]

# @lc code=end
